### ISSUE TITLE
Nullness issue - Spurious nullness warning with flexible type

### ISSUE THREAD

**Issue by glchapman** [dotnet/fsharp#18056]
### Issue description

I get a nullness warning with the following:
```
let dispose (x: IDisposable | null) : unit =
    match x with
    | null -> ()
    | d -> d.Dispose()

let useThing (thing: #IDisposable) =
    try
        printfn "%O" thing
    finally
        dispose thing // warning generated
```

Since the dispose function allows null, I do not think a warning should be generated. No warning is generated for a non-flexible type, e.g.:
```
let useThing (thing: IDisposable) =
    try
        printfn "%O" thing
    finally
        dispose thing // no warning generated 
```

### Choose one or more from the following categories of impact

- [x] Unexpected nullness warning (false positive in nullness checking, code uses --checknulls and langversion:preview).
- [ ] Missing nullness warning in a case which can produce nulls (false negative, code uses --checknulls and langversion:preview).
- [ ] Breaking change related to older `null` constructs in code not using the checknulls switch.
- [ ] Breaking change related to generic code and explicit type constraints (`null`, `not null`).
- [ ] Type inference issue (i.e. code worked without type annotations before, and applying the --checknulls enforces type annotations).
- [ ] C#/F# interop issue related to nullness metadata.
- [ ] Other (none of the categories above apply).

### Operating System

Windows (Default)

### What .NET runtime/SDK kind are you seeing the issue on

.NET SDK (.NET Core, .NET 5+)

### .NET Runtime/SDK version

net9.0

**Comment by glchapman** [comment]
I just reread the section "Inference in generic code" from Nullable Reference Types in F# 9. I thought perhaps it meant that the following should work without a warning, but a warning is still generated
```
let useThing<'d when 'd: not struct and 'd :> IDisposable> (thing: 'd) =
    try
        printfn "%O" thing
    finally
        dispose thing // warning generated
```

**Comment by T-Gro** [comment]
Could you please also include which warning this produces incl. the range (= coordinates in code) for it?

**Comment by glchapman** [comment]
The warning is:
```
Program.fs(12,17): warning FS3261: Nullness warning: The types 'IDisposable' and 'IDisposable | null' do not have equivalent nullability.
```
In VSCode, `thing` in the last line (`dispose thing`) is highlighted.

**Comment by cataggar** [comment]
I agree. I would be great is `use` could handle `IDispoable | null` without a warning.

![Image](https://github.com/user-attachments/assets/43c792b4-9a9b-429e-8886-0283f108ea55)

This is a example using Microsoft.Extensions.Logging [LoggerExtensions.BeginScope(ILogger, String, Object[]) Method](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.beginscope?view=net-9.0-pp).

Is there a way to disable the warning for just a single line? This disables the warning for the entire file:
``` fs
#nowarn "3261" // use IDisposable | null
```

**Comment by baronfel** [comment]
This pattern is also very prevalent when using System.Diagnostics.Activity and related types. It's annoying to deal with and is one of the reasons that @TheAngryByrd made [some helper members](https://github.com/TheAngryByrd/FsOpenTelemetry/blob/master/src/FsOpenTelemetry/FsOpenTelemetry.fs#L474-L582) to paper over the experience.

**Comment by cataggar** [comment]
I'm confirming that I am able to use `Unchecked.nonNull` to make the compiler happy:

```
use scope = this.BeginScope props |> Unchecked.nonNull
```

This will work until `use` support `IDisposable | null` too.
